File: TCPIP/src/49/so_linger_basic_configuration.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <errno.h>

// Demonstrates three different SO_LINGER configurations

void set_linger_default(int sockfd) {
    struct linger sl;
    sl.l_onoff = 0;  // Disable linger
    sl.l_linger = 0;
    
    if (setsockopt(sockfd, SOL_SOCKET, SO_LINGER, &sl, sizeof(sl)) < 0) {
        perror("setsockopt SO_LINGER (default)");
    } else {
        printf("SO_LINGER: Default behavior (l_onoff=0)\n");
        printf("  - close() returns immediately\n");
        printf("  - TCP sends data in background\n");
        printf("  - Normal TIME_WAIT state\n\n");
    }
}

void set_linger_graceful(int sockfd, int timeout_seconds) {
    struct linger sl;
    sl.l_onoff = 1;  // Enable linger
    sl.l_linger = timeout_seconds;
    
    if (setsockopt(sockfd, SOL_SOCKET, SO_LINGER, &sl, sizeof(sl)) < 0) {
        perror("setsockopt SO_LINGER (graceful)");
    } else {
        printf("SO_LINGER: Graceful linger (l_onoff=1, l_linger=%d)\n", 
               timeout_seconds);
        printf("  - close() blocks up to %d seconds\n", timeout_seconds);
        printf("  - Waits for data to be sent and ACKed\n");
        printf("  - Returns -1 on timeout\n\n");
    }
}

void set_linger_abortive(int sockfd) {
    struct linger sl;
    sl.l_onoff = 1;  // Enable linger
    sl.l_linger = 0; // Zero timeout = abortive close
    
    if (setsockopt(sockfd, SOL_SOCKET, SO_LINGER, &sl, sizeof(sl)) < 0) {
        perror("setsockopt SO_LINGER (abortive)");
    } else {
        printf("SO_LINGER: Abortive close (l_onoff=1, l_linger=0)\n");
        printf("  - Sends RST instead of FIN\n");
        printf("  - Discards pending data\n");
        printf("  - Avoids TIME_WAIT state\n");
        printf("  - WARNING: Can lose data!\n\n");
    }
}

int main() {
    int sockfd = socket(AF_INET, SOCK_STREAM, 0);
    if (sockfd < 0) {
        perror("socket");
        return 1;
    }
    
    printf("=== SO_LINGER Configuration Examples ===\n\n");
    
    // Demonstrate each configuration
    set_linger_default(sockfd);
    set_linger_graceful(sockfd, 10);
    set_linger_abortive(sockfd);
    
    // Query current SO_LINGER setting
    struct linger current_linger;
    socklen_t optlen = sizeof(current_linger);
    
    if (getsockopt(sockfd, SOL_SOCKET, SO_LINGER, 
                   &current_linger, &optlen) == 0) {
        printf("Current SO_LINGER setting:\n");
        printf("  l_onoff  = %d\n", current_linger.l_onoff);
        printf("  l_linger = %d seconds\n", current_linger.l_linger);
    }
    
    close(sockfd);
    return 0;
}

File: TCPIP/src/49/so_linger_basic_configuration.rs
use std::net::TcpStream;
use std::os::fd::AsRawFd;
use std::time::Duration;

// Import libc for direct socket option manipulation
extern crate libc;

#[derive(Debug)]
enum LingerMode {
    Default,
    Graceful(Duration),
    Abortive,
}

struct SocketLinger {
    stream: TcpStream,
}

impl SocketLinger {
    fn new(stream: TcpStream) -> Self {
        Self { stream }
    }
    
    fn set_linger(&self, mode: LingerMode) -> Result<(), std::io::Error> {
        let fd = self.stream.as_raw_fd();
        
        let linger_struct = match mode {
            LingerMode::Default => {
                println!("Setting DEFAULT linger mode:");
                println!("  - close() returns immediately");
                println!("  - TCP sends data in background");
                println!("  - Normal TIME_WAIT state");
                
                libc::linger {
                    l_onoff: 0,
                    l_linger: 0,
                }
            }
            LingerMode::Graceful(timeout) => {
                let seconds = timeout.as_secs() as i32;
                println!("Setting GRACEFUL linger mode:");
                println!("  - close() blocks up to {} seconds", seconds);
                println!("  - Waits for data to be sent and ACKed");
                println!("  - Returns error on timeout");
                
                libc::linger {
                    l_onoff: 1,
                    l_linger: seconds,
                }
            }
            LingerMode::Abortive => {
                println!("Setting ABORTIVE linger mode:");
                println!("  - Sends RST instead of FIN");
                println!("  - Discards pending data");
                println!("  - Avoids TIME_WAIT state");
                println!("  - WARNING: Can lose data!");
                
                libc::linger {
                    l_onoff: 1,
                    l_linger: 0,
                }
            }
        };
        
        unsafe {
            let result = libc::setsockopt(
                fd,
                libc::SOL_SOCKET,
                libc::SO_LINGER,
                &linger_struct as *const _ as *const libc::c_void,
                std::mem::size_of::<libc::linger>() as libc::socklen_t,
            );
            
            if result < 0 {
                return Err(std::io::Error::last_os_error());
            }
        }
        
        Ok(())
    }
    
    fn get_linger(&self) -> Result<(bool, i32), std::io::Error> {
        let fd = self.stream.as_raw_fd();
        let mut linger_struct = libc::linger {
            l_onoff: 0,
            l_linger: 0,
        };
        let mut len = std::mem::size_of::<libc::linger>() as libc::socklen_t;
        
        unsafe {
            let result = libc::getsockopt(
                fd,
                libc::SOL_SOCKET,
                libc::SO_LINGER,
                &mut linger_struct as *mut _ as *mut libc::c_void,
                &mut len,
            );
            
            if result < 0 {
                return Err(std::io::Error::last_os_error());
            }
        }
        
        Ok((linger_struct.l_onoff != 0, linger_struct.l_linger))
    }
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    println!("=== SO_LINGER Configuration Examples ===\n");
    
    // Create a socket (not connected, just for demonstration)
    let stream = TcpStream::connect("127.0.0.1:1")
        .or_else(|_| {
            // If connection fails, create a socket anyway for demonstration
            // This is just for showing the API
            use std::net::{SocketAddr, TcpListener};
            let listener = TcpListener::bind("127.0.0.1:0")?;
            let addr = listener.local_addr()?;
            
            // Create a connection to ourselves
            let handle = std::thread::spawn(move || {
                listener.accept().ok()
            });
            
            let stream = TcpStream::connect(addr)?;
            handle.join().ok();
            Ok(stream)
        })?;
    
    let socket = SocketLinger::new(stream);
    
    // Demonstrate DEFAULT mode
    println!("\n--- Configuration 1 ---");
    socket.set_linger(LingerMode::Default)?;
    let (enabled, timeout) = socket.get_linger()?;
    println!("Current: l_onoff={}, l_linger={}\n", enabled, timeout);
    
    // Demonstrate GRACEFUL mode
    println!("--- Configuration 2 ---");
    socket.set_linger(LingerMode::Graceful(Duration::from_secs(10)))?;
    let (enabled, timeout) = socket.get_linger()?;
    println!("Current: l_onoff={}, l_linger={}\n", enabled, timeout);
    
    // Demonstrate ABORTIVE mode
    println!("--- Configuration 3 ---");
    socket.set_linger(LingerMode::Abortive)?;
    let (enabled, timeout) = socket.get_linger()?;
    println!("Current: l_onoff={}, l_linger={}\n", enabled, timeout);
    
    println!("\n=== Recommendations ===");
    println!("• Use DEFAULT for most applications");
    println!("• Use GRACEFUL when data delivery is critical");
    println!("• Use ABORTIVE only in special cases (testing, emergency shutdown)");
    
    Ok(())
}